src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs

```diff
@@ -587,7 +587,7 @@ private void AddLowercaseRange(char chMin, char chMax, CultureInfo culture)
 
             for (i = 0, iMax = s_lcTable.Length; i < iMax;)
             {
-                iMid = (i + iMax) / 2;
+                iMid = (i & iMax) + ((i ^ iMax) >> 1); // Faster than (i + iMax) / 2;
```

Comment:
Interesting. I am happy to revert it as certainly it's less readable. @AndyAyersMS thoughts?

Reply:
[This](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACBmABJgJlwGFcBvAWAChdb88BLAOzlwHkAnBgc2cQBsAFM1YMYuERICyiMAEpy1OsvwB2XMNwBqabIUB6XAQDcSugF8ztHBJa4AcgFMA7sLtjbomfMU0V19U0AMl0fHUFNAD1QhQA+WNwARjlTP1pLKnMgA=) is probably closer to what RyuJit does now.

So the "optimized" sequence is potentially faster  as it has a shallower dependence chain. Whether this win shows up at an actual use depends on what else is going on nearby.

I would suggest leaving leave the simple `/ 2` as is and open an issue over in CoreCLR to have somebody on the jit team look at alternate expansion strategies.

Reply:
This is not a valid transformation. For instance, those two expressions give different results for (-1, 0) and (1, `int.MaxValue`) pairs.  Moreover, `(i + iMax) / 2` evaluates to a negative index in the case of overflow.  Why don't we use `(int)(i + (uint)(iMax - i) / 2)`, which is less fancy and more efficient? See on [sharplab.io](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAmABAYVwG8BYAKF0t0wEYJcBLAOzlwHkAnBgc2cQBsAFM1YMYjFowCyiMAEpi5KsuoB2XMNwBqabIUB6XNgDcSqgF8zlWvRG4AYoiYBjAJ7DJYiaJnzFFFWt1TQAyXT8dQU0APXCFAD543Bo5UwDKS3TqOm9cAGUAV25uAFNkOBKAEw9RcTsGXwVSLOVMYJE5TUiCjuFfXABaRgMjVKtcTPMgA=).

Reply:
Is it not a valid transformation if the values are assumed non negative, as when used in binary search?

Reply:
@danmosemsft For non-negative numbers it is not a valid transformation if there is an overflow.  Consider `1` and `int.MaxValue`: `(i + iMax) / 2` evaluates to `-1073741824`, while `(i & iMax) + ((i ^ iMax) >> 1)` evaluates to `+1073741824`.  Note that the real issue here, both for performance and correctness, is using a signed division instead of unsigned one.  The easiest fix is actually to use an unsigned division: `(int)((uint)(i + iMax) / 2)` (disregard my previous suggestion).  See on [sharplab.io](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAmABAYVwG8BYAKF0t0wEYJcBLAOzlwHkAnBgc2cQBsAFM1YMYjFowCyiMAEpi5KsuoB2XMNwBqabIUB6XNgDcSqgF8zlWvRG4AYoiYBjAJ7DJYiaJnzFFFWt1TQAyXT8dQU0APXCFAD543Bo5UwDKS3TqOm9cAGUAV25uAFNkOBKAEw9RcTsGXwVSLOVMYJE5KIKOzR0GvVxDbFSrXEzzIA).

Reply:
@AntonLapounov right, but in your example `+1073741824` is presumably the desirable result.

I agree your way seems better. I had not thought that the uint cast could "recover" from the overflow after the addition.

Reply:
@danmosemsft Yes, that is the desirable result, but my point was the JIT would never be able to make this or similar transformation, because those expressions are not equivalent.  Actually, `(x & y) + ((x ^ y) >> 1)` gives the correct result for all possible `int` number pairs (positive, zero, or negative), while `(int)((uint)(x + y) / 2)` gives the correct result for non-negative `x`, `y` only.  If you rewrite the latter expression as `(int)(((uint)x + (uint)y) / 2)` (which is equivalent in the unchecked context), it may be easier to see why it "recovers" from the overflow.